### PR TITLE
feat(metrics): add a classification metrics module

### PR DESCRIPTION
**Pull request by IndexSeek**
I wanted to start making some progress on supporting a new `ibis_ml.metrics`
module might look like with the first pass to support classification metrics. 

In this initial pass, the functions accept the `y_pred` and `y_true` as Ibis table
expression columns.

## A few thoughts
- Should we go ahead and try this with regression metrics? Or would it be okay
to start with classification only as an "experimental" type of op.
- Do we need a "metrics" folder that contains "regression" and "classification" files?
- Is there a more effective way to test this than to compare it to scikit-learns metrics?

## What could go wrong?

I anticipate the inputs are coming from the same table expression, although testing
this behavior (or at least documenting) with what happens with input from two different
table expressions would be necessary. 

Resolves (partially) #174 

Edit: As far as testing goes, I wonder if more precision might be necessary.

### CODE REVIEW

**Comment by deepyaman** [comment]
Sorry, I didn't see this earlier; will try to take a look soon!

File: ibis_ml/metrics.py [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+
+    Returns
+    -------
+    float
+        The accuracy score, representing the fraction of correct predictions.
+
+    Examples
+    --------
+    >>> from ibis_ml.metrics import accuracy_score
+    >>> import ibis
```
Comment (deepyaman) [inline ibis_ml/metrics.py:22]: Nit: Can you put the `import ibis` above `from ibis_ml`? Mental `isort`ing. :P

File: ibis_ml/metrics.py [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+    0.5714285714285714
+    """
+    true_positive = (y_true & y_pred).sum()
+    actual_positive = y_true.sum()
+    return (true_positive / actual_positive).to_pyarrow().as_py()
+
+
+def f1_score(y_true: dt.Integer, y_pred: dt.Integer) -> float:
+    """Calculate the F1 score of predicted values against true values.
+    Parameters
```
Comment (deepyaman) [inline ibis_ml/metrics.py:105]:
```suggestion

    Parameters
```

File: ibis_ml/metrics.py [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+    0.6666666666666666
+    """
+    true_positive = (y_true & y_pred).sum()
+    predicted_positive = y_pred.sum()
+    return (true_positive / predicted_positive).to_pyarrow().as_py()
+
+
+def recall_score(y_true: dt.Integer, y_pred: dt.Integer) -> float:
+    """Calculate the recall score of predicted values against true values.
+    Parameters
```
Comment (deepyaman) [inline ibis_ml/metrics.py:72]:
```suggestion

    Parameters
```

File: tests/test_metrics.py [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+    return ibis.memtable(
+        {
+            "id": range(1, 13),
+            "actual": [1, 0, 0, 1, 1, 0, 0, 0, 1, 1, 1, 1],
+            "prediction": [1, 0, 0, 1, 0, 0, 1, 1, 0, 1, 0, 1],
+        }
+    )
+
+
+def test_accuracy_score(results_table):
```
Comment (deepyaman) [inline tests/test_metrics.py:22]: Nit: I think it would be clean to just have one parametrized test that takes a `metric` parameters. Then you can just `import ibis_ml.metrics` and `import sklearn.metrics` and `getattr(ibis_ml.metrics, metric)`, etc.

**Review by deepyaman — CHANGES_REQUESTED** [review]
One functional request to not force compute; rest are quick nits!

File: tests/test_metrics.py (hunk 4 above)
Comment (IndexSeek) [reply tests/test_metrics.py]: I implemented this in https://github.com/ibis-project/ibis-ml/pull/176/commits/959b45d0f78b8bedf5888ddb6f6cb10f50633bb0, this was an excellent idea and did tidy up things nicely. The previous way I was prefixing "sk_" did not feel great.

File: ibis_ml/metrics.py (hunk 1 above)
Comment (IndexSeek) [reply ibis_ml/metrics.py:21]: Good catch! This is fixed now in the four functions.

File: ibis_ml/metrics.py [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+    ...     {
+    ...         "id": range(1, 13),
+    ...         "actual": [1, 0, 0, 1, 1, 0, 0, 0, 1, 1, 1, 1],
+    ...         "prediction": [1, 0, 0, 1, 0, 0, 1, 1, 0, 1, 0, 1],
+    ...     }
+    ... )
+    >>> accuracy_score(t.actual, t.prediction)
+    0.5833333333333334
+    """
+    return (y_true == y_pred).mean().to_pyarrow().as_py()
```
[the remark this answers is not in the document]
Comment (IndexSeek) [reply ibis_ml/metrics.py]: I agree with this, it does seem strange to mix and match eager and lazy evaluation. Keeping it lazy here seems more aligned with everything else. 

I would anticipate users possibly needing to do something with thes values, like logging the metrics as part of a model's metadata. But it isn't too difficult to evaluate the expression.

**Comment by IndexSeek** [comment]
Thanks for your review and the suggestions. I think these were great and have added them all now. 

I'm not sure what's going on with the build-docs check here, it seems the last time it succeeded it was using scikit-learn==1.5.2 and now it's using scikit-learn==1.6.0.

**Comment by deepyaman** [comment]
> I'm not sure what's going on with the build-docs check here, it seems the last time it succeeded it was using scikit-learn==1.5.2 and now it's using scikit-learn==1.6.0.

Seems `__sklearn_tags__` is new from https://github.com/scikit-learn/scikit-learn/pull/29677; I'll have to look into it. Feel free to pin `scikit-learn <1.6.0` for now in https://github.com/ibis-project/ibis-ml/blob/main/pyproject.toml#L23.

**Comment by IndexSeek** [comment]
It's pinned now at <1.6.0 for the "doc" dependencies and the check is passing! ✅

**Review by deepyaman — APPROVED** [review]
LGTM!
